### PR TITLE
MeasureDef and GroupDef Refactor

### PR DESCRIPTION
**Pull request by Capt-Mac**
GroupDef now holds definition for populationBasis, MeasureScoring, and ImprovementNotation

### CODE REVIEW

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureReportBuilder.java [hunk 1]
```diff
@@ -286,7 +286,7 @@ protected void buildPopulation(
         reportPopulation.setCode(measurePopulation.getCode());
         reportPopulation.setId(measurePopulation.getId());
 
-        if (measureDef.isBooleanBasis()) {
+        if (measureDef.groups().get(0).isBooleanBasis()) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureReportBuilder.java]: What happens if groups is empty?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java [hunk 2]
```diff
@@ -231,27 +221,14 @@ private MeasureScoring getMeasureScoring(Measure measure) {
...
-        }
-        // otherwise return measureLevelScoring
-        return measureLevelScoring;
-    }
-
-    private boolean isIncreaseImprovementNotation(CodeableConcept improvementNotationValue) {
-        validateImprovementNotationCode(improvementNotationValue);
-        return improvementNotationValue.hasCoding(
-                MEASUREREPORT_IMPROVEMENT_NOTATION_SYSTEM, IMPROVEMENT_NOTATION_SYSTEM_INCREASE);
+    private MeasureScoring getMeasureScoring(Measure measure) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java:224]: General suggestion:  Mark method parameters you expect to be null with @Nullable.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java [hunk 3]
```diff
@@ -262,27 +239,102 @@ private void validateImprovementNotationCode(CodeableConcept improvementNotation
...
+            return codeDef;
         }
+        return null;
+    }
+
+    public CodeDef getGroupImpNotation(MeasureGroupComponent group) {
+        var ext = group.getExtensionByUrl(MEASUREREPORT_IMPROVEMENT_NOTATION_EXTENSION);
+        if (ext != null) {
+            var value = ext.getValue();
+            assert value instanceof CodeableConcept;
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java:275]: assert is disabled by default by Java and must be enabled explicitly.  I recommend either conditional logic and throwing an explicit Exception, or using a library like Guava's Preconditions.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureReportBuilder.java (hunk 1 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureReportBuilder.java]: this method is not used and was removed

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java [hunk 4]
```diff
@@ -262,27 +239,102 @@ private void validateImprovementNotationCode(CodeableConcept improvementNotation
...
+        return null;
+    }
+
+    private MeasureScoring getScoringDef(MeasureScoring measureScoring, MeasureScoring groupScoring) {
+        if (groupScoring == null && measureScoring == null) {
+            throw new IllegalArgumentException("MeasureScoring must be specified on Group or Measure");
+        }
+        if (groupScoring != null) {
+            return groupScoring;
+        } else {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java:312]: Nitpick:  return short-circuits execution of if, so you don't need else to follow it, just return after the if block.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureReportBuilder.java [hunk 5]
```diff
@@ -460,10 +456,9 @@ protected void addMeasureDescription(MeasureReportGroupComponent reportGroup, Me
         }
     }
 
-    protected void addExtensionImprovementNotation(
-            MeasureReportGroupComponent reportGroup, MeasureDef measureDef, GroupDef groupDef) {
+    protected void addExtensionImprovementNotation(MeasureReportGroupComponent reportGroup, GroupDef groupDef) {
         // if already set on Measure, don't set on groups too
-        if (!measureDef.useMeasureImpNotation()) {
+        if (groupDef.isGroupImprovementNotation()) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureReportBuilder.java:461]: The previous if was a negative condition an this one is positive.  Just wanted to make sure that was your intention.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilderTest.java [hunk 6]
```diff
@@ -50,7 +54,7 @@ void defBuilderBooleanBasis_WithResourceExt() {
         measure.setExtension(Collections.singletonList(RESOURCE_BASIS_EXT));
 
         var def = BUILDER.build(measure);
-        assertFalse(def.isBooleanBasis());
+        assertFalse(def.groups().get(0).isBooleanBasis());
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilderTest.java:57]: Suggestion:  assert that groups.size() > 0 before making these kinds of assertions

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+            CodeableConcept measureImpNotation) {
+
+        R4MeasureDefBuilder defBuilder = new R4MeasureDefBuilder();
+        org.hl7.fhir.r4.model.Measure measure = (org.hl7.fhir.r4.model.Measure)
+                parser.parseResource(MeasureDefBuilderTest.class.getResourceAsStream("TemplateMeasure.json"));
+
+        var group1 = measure.getGroup().stream()
+                .filter(t -> t.getId().equals("group-1"))
+                .findFirst()
+                .get();
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java:60]: Suggestion:  orElseThrow or some clear signal if the groups are empty

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+                MeasureScoring.RATIO,
+                true,
+                "boolean",
+                false,
+                "decrease",
+                MeasureScoring.PROPORTION);
+    }
+
+    @Test
+    void noImprovementNotation() {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java:428]: Great test coverage!

**Review by lukedegruchy — APPROVED** [review]

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureReportBuilder.java (hunk 5 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureReportBuilder.java:461]: good question, luckily these are looking at different references to check. One is Measure level the other group level. this is correct. behavior

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java (hunk 7 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureDefBuilderTest.java:61]: this is a static template for this test, so no risk of not finding group

**Pull request merged**